crates/ide-diagnostics/src/handlers/type_mismatch.rs

```diff
@@ -184,6 +189,86 @@ fn add_missing_ok_or_some(
     Some(())
 }
 
+fn remove_unnecessary_wrapper(
+    ctx: &DiagnosticsContext<'_>,
+    d: &hir::TypeMismatch,
+    expr_ptr: &InFile<AstPtr<ast::Expr>>,
+    acc: &mut Vec<Assist>,
+) -> Option<()> {
+    let root = ctx.sema.db.parse_or_expand(expr_ptr.file_id);
+    let expr = expr_ptr.value.to_node(&root);
+    let expr_range = expr.syntax().text_range();
+    let actual_enum = d.actual.as_adt().and_then(|adt| adt.as_enum())?;
+    let famous_defs = FamousDefs(&ctx.sema, ctx.sema.scope(expr.syntax())?.krate());
+    let file_id = expr_ptr.file_id.original_file(ctx.sema.db);
+
+    let Expr::CallExpr(call_expr) = expr else {
+        return None;
+    };
+
+    let path_str = call_expr.syntax().descendants().find_map(ast::PathExpr::cast)?.to_string();
```

Comment:
We shouldn't do stringly checks, let's inspect the actual ADT type here instead, `d.actual.as_adt()` and then compare that to the relevant famous defs to check which one it is.

Reply:
This gets me the relevant enum type, but I can't seem to figure out how to get which specific was variant constructed. Any idea?